src/main/kotlin/org/rust/lang/core/psi/ProcMacroAttribute.kt

```diff
@@ -0,0 +1,164 @@
+    object Derive : ProcMacroAttribute<Nothing>() {
+        override val attr: Nothing? get() = null
+        override fun toString(): String = "Derive"
+    }
+
+    /**
+     * The item **has** attribute procedural macro call, it is [attr]
+     */
+    data class Attr<T : RsMetaItemPsiOrStub>(
+        override val attr: T,
+        /**
+         * The index of [attr] (starting from 0) in [RsDocAndAttributeOwner.queryAttributes]
+         * of the attribute owner
+         */
+        val index: Int
+    ) : ProcMacroAttribute<T>()
+
+    companion object {
+        /**
+         *
+         * Can't be after derive:
+         *
+         * ```
+         * #[derive(Foo)]
+         * #[foo] // NOT an attribute macro
+         * struct S
+         * ```
+         *
+         * Legacy derive helpers aren't supported
+         * https://github.com/rust-lang/rust/issues/79202
+         */
+        fun <T : RsMetaItemPsiOrStub> getProcMacroAttributeRaw(
+            owner: RsAttrProcMacroOwnerPsiOrStub<T>,
+            stub: RsAttributeOwnerStub? = if (owner is RsDocAndAttributeOwner) owner.attributeStub else owner as RsAttributeOwnerStub,
+            explicitCrate: Crate? = null,
+            explicitCustomAttributes: CustomAttributes? = null,
+        ): ProcMacroAttribute<T> {
+            if (!ProcMacroApplicationService.isEnabled()) return None
+            if (stub != null) {
+                if (!stub.mayHaveCustomAttrs) {
+                    return if (stub.mayHaveCustomDerive) Derive else None
+                }
+            }
+
+            val crate = explicitCrate ?: owner.containingCrate ?: return None
+            val customAttributes = explicitCustomAttributes ?: CustomAttributes.fromCrate(crate)
+
+            owner.getQueryAttributes(crate, stub, fromOuterAttrsOnly = true).metaItems.forEachIndexed { index, meta ->
+                if (meta.name == "derive") return Derive
+                if (RsProcMacroPsiUtil.canBeProcMacroAttributeCallWithoutContextCheck(meta, customAttributes)) {
+                    return Attr(meta, index)
+                }
+            }
+            return None
+        }
+
+        fun getProcMacroAttribute(
+            owner: RsAttrProcMacroOwner,
+            stub: RsAttributeOwnerStub? = owner.attributeStub,
+            explicitCrate: Crate? = null,
+        ): ProcMacroAttribute<RsMetaItem> {
+            return when (val attr = getProcMacroAttributeRaw(owner, stub, explicitCrate)) {
+                Derive -> Derive
+                None -> None
+                is Attr -> {
+                    val props = attr.attr.resolveToProcMacroWithoutPsiUnchecked(checkIsMacroAttr = false)?.props
+                    if (props != null && props.treatAsBuiltinAttr && RsProcMacroPsiUtil.canFallbackItem(owner)) {
+                        None
+                    } else {
+                        attr
+                    }
```

Comment:
Is it true that we return some `Attr` instance even if we failed to resolve proc macro?

Reply:
Absolutely